core/domain/interaction_validation_jobs_one_off.py

```diff
@@ -0,0 +1,278 @@
+    def reduce(key, values):
+        yield (key, values)
+
+
+class ItemSelectionInteractionOneOffJob(jobs.BaseMapReduceOneOffJobManager):
+    """Job that produces a list of (exploration, state) pairs that use the item
+    selection interaction and that have rules that do not match the answer
+    choices. These probably need to be fixed manually.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+
+        exploration = exp_fetchers.get_exploration_from_model(item)
+        for state_name, state in exploration.states.items():
+            if state.interaction.id == 'ItemSelectionInput':
+                choices = (
+                    state.interaction.customization_args['choices']['value'])
+                for group in state.interaction.answer_groups:
+                    for rule_spec in group.rule_specs:
+                        for rule_item in rule_spec.inputs['x']:
+                            if rule_item not in choices:
+                                yield (
+                                    item.id,
+                                    '%s: %s' % (
+                                        state_name.encode('utf-8'),
+                                        rule_item.encode('utf-8')))
+
+    @staticmethod
+    def reduce(key, values):
+        yield (key, values)
+
+
+class InteractionRTECustomizationArgsValidationOneOffJob(
+        jobs.BaseMapReduceOneOffJobManager):
+    """One-off job for validating all the customizations arguments of
+    Rich Text Components.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(item):
+        if item.deleted:
+            return
+        err_dict = {}
+
+        try:
+            exploration = exp_fetchers.get_exploration_from_model(item)
+        except Exception as e:
+            yield (
+                'Error %s when loading exploration'
+                % python_utils.UNICODE(e), [item.id])
+            return
+
+        html_list = exploration.get_all_html_content_strings()
+        err_dict = html_validation_service.validate_customization_args(
+            html_list)
+
+        for key in err_dict:
+            if err_dict[key]:
+                yield ('%s Exp Id: %s' % (key, item.id), err_dict[key])
+
+    @staticmethod
+    def reduce(key, values):
+        final_values = [ast.literal_eval(value) for value in values]
+        # Combine all values from multiple lists into a single list
+        # for that error type.
+        output_values = list(set().union(*final_values))
+        exp_id_index = key.find('Exp Id:')
+        if exp_id_index == -1:
+            yield (key, output_values)
+        else:
+            output_values.append(key[exp_id_index:])
+            yield (key[:exp_id_index - 1], output_values)
```

Comment:
Would we need the state name to debug / actually find an example of the problematic interaction in question? Or is that not needed at this stage?

Reply:
I think not needed at this stage - we get the issues and find the way to migrate and fix and in the end enable the validation before save.